### PR TITLE
Add Markdown passthrough

### PR DESCRIPTION
**Pull request by takluyver**
This is meant as a base class, which I can build on to allow modifying Markdown programmatically without converting it to another format. It's rather clumsy, and I'm not yet sure if this is the route I'll go, but I'm putting it here for ideas and discussion.

### CODE REVIEW

**Comment by mpacer** [comment]
Just to confirm, this would use the internal parsing machinery of mistune to give a data structure that can be directly modified and returned as markdown text? 

This would be useful for a variety of use cases in https://github.com/jupyter/nbconvert, in particular this is exactly what I'm looking for when it comes to modifying header contents to create automatic header ids. 

If this isn't going to be merged into mistune-contrib, can I suggest we port this pr to Nbconvert (perhaps into utils?) for our direct use even if we don't expose it as public API?

**Comment by takluyver** [comment]
Kind of; because of the way mistune works, this doesn't actually give you a data structure representing the parsed document. Instead, it implements a renderer which outputs the document as Markdown. The idea is that if you want to modify something in Markdown, you would subclass `MarkdownTransformer` and override the specific methods to process the pieces you want (e.g. `output_link()`).

I think it would be best for this to be in mistune (or -contrib), because it has a fairly large API surface coupled with mistune. But if @lepture isn't interested, then we can try to maintain this in nbconvert. I seem to remember that the current implementation is a bit rough and ready in places, though.

File: mistune_contrib/passthrough.py [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+        self.inline.setup(self.block.def_links, self.block.def_footnotes)
+
+        results = []
+        while self.pop():
+            results.append(self.tok())
+        return ''.join(results)
+
+    def tok(self):
+        t = self.token['type']
+        #print('block', t, repr(self.token['text']))
```
Comment (lepture) [inline mistune_contrib/passthrough.py]: cleanup this

File: mistune_contrib/passthrough.py [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+        if t.endswith('_start'):
+            t = t[:-6]
+
+        return getattr(self, 'output_%s' % t, self.block_default)()
+
+    def inline_tokens(self, text, rules=None):
+        results = []
+        for t in self.inline.parse(text, rules=rules):
+            meth = getattr(self, 'output_'+t['type'], self.inline_default)
+            #print('inline', t['type'], repr(t['text']))
```
Comment (lepture) [inline mistune_contrib/passthrough.py]: cleanup

File: mistune_contrib/passthrough.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+                    continue
+                return m, key
+            return False  # pragma: no cover
+
+        tokens = []
+
+        self.line_started = False
+        while text:
+            ret = manipulate(text)
+            self.line_started = True
```
Comment (lepture) [inline mistune_contrib/passthrough.py]: line_started is useless in the code. I've removed it in mistune.

File: mistune_contrib/passthrough.py [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+        # sepcial cases
+        if t.endswith('_start'):
+            t = t[:-6]
+
+        return getattr(self, 'output_%s' % t, self.block_default)()
+
+    def inline_tokens(self, text, rules=None):
+        results = []
+        for t in self.inline.parse(text, rules=rules):
+            meth = getattr(self, 'output_'+t['type'], self.inline_default)
```
Comment (lepture) [inline mistune_contrib/passthrough.py:135]: validate syntax with flake8.

**Comment by takluyver** [comment]
This totally dropped off my radar, but I've tried to make the requested changes now.

File: mistune_contrib/passthrough.py (hunk 4 above)
Comment (takluyver) [reply mistune_contrib/passthrough.py:135]: Not sure if you wanted a change on this particular line, but I've made flake8 happy.

**Pull request merged**